test/Z3/Base/Spec.hs

```diff
@@ -180,6 +180,71 @@ spec = around withContext $ do
           Z3.getBoolValue ctx ast
         assert $ x == Just b
 
+    specify "mkAtMost0" $ \ctx ->
+      (do
+        a <- Z3.mkFreshBoolVar ctx "A"
+        b <- Z3.mkFreshBoolVar ctx "B"
+        c <- Z3.mkFreshBoolVar ctx "C"
+        solver <- Z3.mkSolver ctx
+        Z3.solverAssertCnstr ctx solver =<< Z3.mkAtMost ctx [a, b, c] 0
+        (_r, Just model) <- Z3.solverCheckAndGetModel ctx solver
+        mapM (Z3.evalBool ctx model) [a, b, c])
+      `shouldReturn` [Just False, Just False, Just False]
+
+    specify "mkAtMost2" $ \ctx ->
+      (do
+        a <- Z3.mkFreshBoolVar ctx "A"
+        b <- Z3.mkFreshBoolVar ctx "B"
+        c <- Z3.mkFreshBoolVar ctx "C"
+        solver <- Z3.mkSolver ctx
+        Z3.solverAssertCnstr ctx solver =<< Z3.mkAtMost ctx [a, b, c] 2
+        Z3.solverAssertCnstr ctx solver =<< Z3.mkAnd ctx [a, b, c]
+        Z3.solverCheck ctx solver)
+      `shouldReturn` Z3.Unsat
+
+    specify "mkAtMost4" $ \ctx ->
+      (do
+        a <- Z3.mkFreshBoolVar ctx "A"
+        b <- Z3.mkFreshBoolVar ctx "B"
+        c <- Z3.mkFreshBoolVar ctx "C"
+        solver <- Z3.mkSolver ctx
+        Z3.solverAssertCnstr ctx solver =<< Z3.mkAtMost ctx [a, b, c] 4
+        (_r, Just model) <- Z3.solverCheckAndGetModel ctx solver
+        mapM (Z3.evalBool ctx model) [a, b, c])
+      `shouldReturn` [Just False, Just False, Just False]
```

Comment:
This is not necessarily guaranteed, or am I misreading something? Any assignment to the variables will satisfy the constraint, so this test could fail if z3 decides to choose another model.

Reply:
That's correct, should I just check for `Z3.Sat` and not bother with the model or implement `atLeast/Most :: [Maybe Bool] -> Bool` and assert that instead?

Reply:
Whatever you prefer, but in my opinion the second option is a bit more testing Z3 itself so personally I would just check sat in this case.

Reply:
Ok, fixed!